readthedocs/proxito/views/mixins.py

```diff
@@ -169,19 +173,40 @@ def canonical_redirect(self, request, final_project, version_slug, filename):
         """
         Return a redirect to the canonical domain including scheme.
 
-        This is normally used HTTP -> HTTPS redirects or redirects to/from custom domains.
+        The following cases are covered:
+
+        - Redirect a custom from http to https (if supported)
+          http://project.rtd.io/ -> https://project.rtd.io/
+        - Redirect a domain to a canonical domain (http or https).
+          http://project.rtd.io/ -> https://docs.test.com/
+          http://project.rtd.io/foo/bar/ -> https://docs.test.com/foo/bar/
+        - Redirect from a subproject domain to the main domain
+          https://subproject.rtd.io/en/latest/foo -> https://main.rtd.io/projects/subproject/en/latest/foo  # noqa
+          https://subproject.rtd.io/en/latest/foo -> https://docs.test.com/projects/subproject/en/latest/foo  # noqa
         """
-        full_path = request.get_full_path()
-        urlparse_result = urlparse(full_path)
-        to = resolve(
-            project=final_project,
-            version_slug=version_slug,
-            filename=filename,
-            query_params=urlparse_result.query,
-            external=hasattr(request, 'external_domain'),
-        )
+        from_url = request.build_absolute_uri()
+        parsed_from = urlparse(from_url)
 
-        if full_path == to:
+        redirect_type = getattr(request, 'canonicalize', None)
+        if redirect_type == REDIRECT_HTTPS:
+            to = parsed_from._replace(scheme='https').geturl()
+        else:
+            to = resolve(
+                project=final_project,
+                version_slug=version_slug,
+                filename=filename,
+                query_params=parsed_from.query,
+                external=hasattr(request, 'external_domain'),
+            )
+            # When a canonical redirect is done, only change the domain.
+            if redirect_type == REDIRECT_CANONICAL_CNAME:
+                to_parsed = urlparse(to)
```

Comment:
This should be called `parsed_to`, or the other should be `from_parsed` to keep them consistent.